BUG, MAINT: Stop using the error-prone deprecated Py_UNICODE apis

eric-wieser opened the pull request:
These APIs work with either UCS2 or UCS4, depending on the value of `Py_UNICODE_WIDE`.
After python 3.3, there's a better way to handle this type of thing, which means we no longer have to care about this.

Fixes gh-3258
Fixes gh-15363

eric-wieser commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -2644,12 +2652,6 @@ STRING_nonzero (char *ip, PyArrayObject *ap)
-#define PyArray_UCS4_ISSPACE Py_UNICODE_ISSPACE
-#else
-#define PyArray_UCS4_ISSPACE(ch) Py_STRING_ISSPACE((char)ch)
-#endif
```
Pretty sure this was a bug, but constructing a failing string is non-trivial

eric-wieser commented on numpy/core/src/multiarray/buffer.c:
```diff
@@ -832,11 +832,6 @@ gentype_getbuffer(PyObject *self, Py_buffer *view, int flags)
-    if (descr->type_num == NPY_UNICODE) {
-        elsize >>= 1;
-    }
-#endif
```
This code is unreachable for unicode anyway, it seems - but this change still makes it more correct

eric-wieser replied:
Update: by adding the buffer interface to Unicode in the second commit, this became reachable.

eric-wieser commented on numpy/core/src/multiarray/common.c:
```diff
@@ -162,19 +162,16 @@ PyArray_DTypeFromObjectHelper(PyObject *obj, int maxdims,
                 }
-                itemsize = PyUnicode_GetLength(temp);
+                itemsize = PyBytes_Size(temp);
             }
```
This looked plain wrong, so I changed it. No idea how this code is hit

eric-wieser replied:
Do you mean S3 or U3 here?

seberg replied:
Sorry, use `b"as"` above, and it will be "S3" not U3.

eric-wieser replied:
Yeah, you're right - with this change the resulting array is type objects instead of bytes.

Arguably the current behavior is nonsense because coercion of arbitrary objects to bytes is a python 2 thing, python 3 should be pushing the user to pick encodings.

I'll roll back these lines.

eric-wieser replied:
Rolled back this change, and added some comments to make the original intent a little clearer.

seberg replied:
Thanks, I obviously agree that this only makes sense on Python 2, but OTOH, the switch is much less clear in the NumPy world...

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -390,6 +390,10 @@ format_@name@(@type@ val, npy_bool scientific,
+ * 
+ * FIXME:
+ *   is this really a good idea?
+ *   stop using Py_UNICODE here.
```
See #15477

Zac-HD commented:
Is this the patch you mentioned for #15363?  If so the example there would be a nice regression test :smile:

eric-wieser commented:
It is indeed - unfortunately, it segfaults and I don't yet know why

eric-wieser commented:
It's not the cause of the segfault, but it turns out that using `PyUnicode_New(size,  0x10ffff)` is not legal unless I actually write a character greater than `0x10000` to the underlying buffer. Otherwise `str.__eq__` behaves incorrectly.

charris commented:
Welcome to the swamp :)

eric-wieser committed:
ENH: Implement the buffer protocol on numpy str_ scalars

This eliminates the need for special casing in `np.generic.__reduce__`

eric-wieser commented:
@mattip, any idea why this fails on PyPI? https://dev.azure.com/numpy/numpy/_build/results?buildId=8052&view=logs&j=cb7817a0-aed9-5b1b-ffdb-285e933f569f&t=713c0b5c-565d-5b6b-cd58-66e4f64bd786&l=3737

It looks as though pypi has an incorrect implementation of `PyUnicode_FromKindAndData`

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -7854,6 +7854,37 @@ def test_null_inside_bstring_array_is_truthy(self):
+    """
+    @pytest.mark.xfail(IS_PYPY, reason=
+        "PyPy overloads this, and uses `.decode('utf-32-le')` without "
+        "`errors='surrogatepass'`")
```
https://bitbucket.org/pypy/pypy/issues/3165

mattip replied:
Is CPython allowing you to create a unicode character `c` where `ord(c) >= sys.maxunicode`?

eric-wieser replied:
Good point, I should include a test for that.

eric-wieser replied:
With this patch:
```
>>> x = np.array(bytes([255]*4)).view('U1')
>>> x[()]
SystemError: invalid maximum character passed to PyUnicode_New
```
Before this patch:
```
>>> x = np.array(bytes([255]*4)).view('U1')
>>> x[()]
UnicodeDecodeError: 'utf-32-le' codec can't decode bytes in position 0-3: code point not in range(0x110000)
```

eric-wieser replied:
We could revert back to using `PyUnicode_DecodeUTF32(..., "surrogatepass")`, but that would be a lot slower, for very little value.

seberg replied:
Sounds strange enough that I would not worry (I did follow it too closely). If only the error type changes (even if SystemError is a bit annoying), I would not worry. If there are more corner case changes along strange things like surrogates, lets just mention it in the release notes.

eric-wieser replied:
Yeah, it's only the error type that changes. We could catch the exception and replace it if we wanted - or perhaps just change `PyUnicode_FromKindAndData` to raise `UnicodeDecodeError` if it discovers a codepoint that is too large.

mattip replied:
> This fails on PyPy

Should now be fixed since the PyPy CI run uses the latest nightly, and the bug was fixed there.

eric-wieser replied:
I removed the commit that added the xfail, let's see...

eric-wieser replied:
That did the trick, thanks!

eric-wieser commented:
@Zac-HD: Regression test added

mattip commented:
LGTM

eric-wieser commented:
I have two possible worries with this patch:

* `np.str_` scalars may get a little more memory intensive.
* Making `np.str_` support the buffer interface might allow silent conversion to `bytes` where previously none was allowed. I haven't found a case of this yet.

If reviewers can constructs possible corner cases for this second point, it would be helpful.

eric-wieser commented:
> I haven't found a case of this yet.

Here's one such case, which used to error, but now assigns the UCS4 codepoints directly:
```python
>>> a = np.zeros(1, 'V8')
>>> a[:] = 'te'
>>> a
array([b'\x74\x00\x00\x00\x65\x00\x00\x00'], dtype='|V8')
```
I don't think this is any weirder than the behavior of `a[:] = 1.0`, which assigns the float bits to the void element.

eric-wieser commented:
And another one:
```
>>> a = np.str_("test")
>>> b"%s" % a
 b't\x00\x00\x00e\x00\x00\x00s\x00\x00\x00t\x00\x00\x00'
```

seberg commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -476,18 +472,29 @@ UNICODE_setitem(PyObject *op, void *ov, void *vap)
-    if ((ptr == NULL) || (PyErr_Occurred())) {
+
+    /* needs to be in canonical form for PyUnicode_GET_LENGTH below */
+    if (PyUnicode_READY(temp) < 0) {
```
I was curious if it is really necessary, but the cpython code is not that quickly figured out :). Nvm, lets just have it there in any case.

eric-wieser replied:
I could switch to the non-macro form of `GetLength` here, this is a remnant of a broken strategy.

eric-wieser replied:
Made that switch

seberg replied:
Not sure what is better, but whatever, this is good :).

seberg reviewed (commented):
The changes look good to me, although I did not look super closely. I think the Length discovery in the dtype discovery code was correct before. Was worried for a second about the delayed initialization, but it seems fine.
As for supporting the buffer interface... Yeah, it should mostly be strange with respect to void (and maybe including buffered) datatypes. But overall, it should be strange enough that it does not matter. I suppose we could mention it in the release notes.

eric-wieser committed:
BUG, MAINT: Stop using the error-prone deprecated Py_UNICODE apis

These APIs work with either UCS2 or UCS4, depending on the value of `Py_UNICODE_WIDE`.
After python 3.3, there's a better way to handle this type of thing, which means we no longer have to care about this.

Fixes gh-3258
Fixes gh-15363

eric-wieser committed:
MAINT,TST: Tidy test_datetime_memoryview a little

eric-wieser commented:
Updated with a test of `memoryview(np.str_)`

eric-wieser commented on numpy/core/defchararray.py:
```diff
@@ -2679,25 +2679,6 @@ class adds the following functionality:
-            obj = bytes(obj)
-
         return chararray(shape, itemsize=itemsize, unicode=unicode,
                          buffer=obj, order=order)
```
This code never made sense in the first place, as `chararray.__new__` has an identity crisis over whether it's trying to be `np.ndarray.__new__` or `np.array`, and accepts `str` objects in place of the `buffer`.

mattip commented:
The test failure is the windows heisenbug. I will put this in soon if there are no objections.

hameerabbasi commented on numpy/core/include/numpy/arrayscalars.h:
```diff
@@ -135,7 +135,13 @@ typedef struct {
+        /* note that the PyObject_HEAD macro lives right here */
+        PyUnicodeObject base;
+        Py_UCS4 *obval;
+} PyUnicodeScalarObject;
```
Do we need to worry about ABI changes here?

mattip replied:
We could check sizeof(PyUnicodeObject) == sizeof(PyUnicodeScalarObject.base) at import

Also this should hit the release note.

eric-wieser replied:
@mattip, your comment suggests an attempt at resilience to cpython ABI changes, which is something we don't care about because we already require recompilation against every version of python.

What hameer is point out is that anyone allocating subclassing our scalars in C (why would anyone do that), or allocating a PyUnicodeScalarObject on the stack (impossible to do without poking at CPython internals) will notice that our struct changed size.

hameerabbasi replied:
It was mentioned in the meeting that there's an API change here if you access the slots as well.

eric-wieser replied:
The slots were all private CPython implementation details, so anyone accessing them was already fending for themselves

seberg replied:
Ah, so API is no issue, and I fully agree that the ABI (size change) is absurd to give much heed to... We could note it in the release notes, but thinking about it more, I am not even sure that makes any sense.

hameerabbasi replied:
If they’re private, I’m okay with putting this in.

seberg commented:
Thanks for the threat Matti, and thanks Eric. Had another glance over and it looks good to me, so putting it in. The test takes around half a second on my computer. A bit slow, but probably fine (and easily changed later).
